### PR TITLE
Add get_documentation() to EnumMessage.

### PR DESCRIPTION
**Pull request by orenbenkiki**
Implement #203.

### CODE REVIEW

**Comment by Peternator7** [comment]
Thanks for the PR. Just a couple of minor tweaks I think.

I'm a little worried that people will mis-use this feature, and negatively impact the quaility of their own documentation, but I also see the practical benefit of something like this so I think it's okay.

**Comment by orenbenkiki** [comment]
What are the minor tweaks you'd like to see?

**Review by Peternator7 — CHANGES_REQUESTED** [review]

**Comment by Peternator7** [comment]
I forgot to hit submit review :( sorry about that

File: strum_macros/src/helpers/metadata.rs [hunk 1]
```diff
@@ -270,7 +274,13 @@ pub trait VariantExt {
 
 impl VariantExt for Variant {
     fn get_metadata(&self) -> syn::Result<Vec<VariantMeta>> {
-        get_metadata_inner("strum", &self.attrs)
+        let result = get_metadata_inner("strum", &self.attrs)?;
+        self.attrs.iter().filter(|attr| attr.path.is_ident("doc")).try_fold(result, |mut vec, attr| {
```
[the remark this answers is not in the document]
Comment (orenbenkiki) [reply strum_macros/src/helpers/metadata.rs:278]: Done.

File: strum_macros/src/macros/enum_messages.rs [hunk 2]
```diff
@@ -65,10 +67,17 @@ pub fn enum_message_inner(ast: &DeriveInput) -> syn::Result<TokenStream> {
...
+            // Push the detailed message.
             detailed_arms
                 .push(quote! { &#name::#ident #params => ::core::option::Option::Some(#msg) });
         }
+
+        if !documentation.is_empty() {
+            let params = params.clone();
+            // Push the documentation.
+            documentation_arms
+                .push(quote! { &#name::#ident #params => ::core::option::Option::Some(concat!(#( #documentation ),*).trim()) });
```
[the remark this answers is not in the document]
Comment (orenbenkiki) [reply strum_macros/src/macros/enum_messages.rs:79]: To make this friendlier to markup, I remove the leading space from each line, and only if there are multiple lines, I suffix each one with a new line when I concatenate them. This way a single-line comment becomes a simple string (w/o a final newline) but a multi-line comment preserves the indentation structure and the final newline in each line.

File: strum_tests/tests/enum_message.rs [hunk 3]
```diff
@@ -2,14 +2,21 @@ use strum::EnumMessage;
 
 #[derive(Debug, Eq, PartialEq, EnumMessage)]
 enum Pets {
+    // I'm a silent dog.
```
[the remark this answers is not in the document]
Comment (orenbenkiki) [reply strum_tests/tests/enum_message.rs:5]: Done.

File: strum_tests/tests/enum_message.rs [hunk 4]
```diff
@@ -2,14 +2,21 @@ use strum::EnumMessage;
...
+    /// And fish.
     #[strum(message = "I'm a cat")]
     #[strum(detailed_message = "I'm a very exquisite striped cat")]
     Cat,
+    /// I'm a fish.
     #[strum(detailed_message = "My fish is named Charles McFish")]
     Fish,
+    /// I'm a bird.
     Bird,
+    /// I'm disabled.
```
[the remark this answers is not in the document]
Comment (orenbenkiki) [reply strum_tests/tests/enum_message.rs:19]: Done.

File: strum_macros/Cargo.toml [hunk 5]
```diff
@@ -26,4 +26,4 @@ rustversion = "1.0"
 syn = { version = "1.0", features = ["parsing", "extra-traits"] }
 
 [dev-dependencies]
-strum = "0.20"
+strum = { path = "../strum" }
```
[the remark this answers is not in the document]
Comment (orenbenkiki) [reply strum_macros/Cargo.toml:29]: This is consistent with the way that `strum_tests` refers to the parent crate - I took it from there. So the answer should be "yes" for tests and I _assume_ also for docs? Is there a way to test this?

**Comment by orenbenkiki** [comment]
Did a pass (trying) to address the review issues.

**Comment by orenbenkiki** [comment]
Ok, checks pass now.

File: strum_macros/Cargo.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (Peternator7) [reply strum_macros/Cargo.toml:29]: I think it should be fine. My concern was that strum_tests isn't published to crates.io so it's behavior might not be correct for a published library. However, [this page](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#development-dependencies) suggests that it doesn't make a big difference

**Review by Peternator7 — APPROVED** [review]

**Comment by Peternator7** [comment]
Released as part of `0.24`

**Pull request merged**